nixos/modules/services/cluster/kubernetes/default.nix

```diff
@@ -103,12 +103,18 @@ let
     keyFile = mkDefault cfg.kubeconfig.keyFile;
   };
 
-  cniConfig = pkgs.buildEnv {
-    name = "kubernetes-cni-config";
-    paths = imap (i: entry:
-      pkgs.writeTextDir "${toString (10+i)}-${entry.type}.conf" (builtins.toJSON entry)
-    ) cfg.kubelet.cni.config;
-  };
+  cniConfig =
+    if cfg.kubelet.cni.config != [] && !(isNull cfg.kubelet.cni.configDir) then
+      throw "Verbatim CNI-config and CNI configDir cannot both be set."
+    else if !(isNull cfg.kubelet.cni.configDir) then
+      cfg.kubelet.cni.configDir
+    else
+      (pkgs.buildEnv {
+        name = "kubernetes-cni-config";
+        paths = imap (i: entry:
+          pkgs.writeTextDir "${toString (10+i)}-${entry.type}.conf" (builtins.toJSON entry)
```

Comment:
Can we add something like `autogenerated_by_NixOS_module` to the filename? (Also fine for a separate PR)

Reply:
I don't know what other modules do when writing files to the store? Is it standard practice to write such header in all NixOS module generated files? In any case, yes, let's consider it for a separate PR :)

Reply:
I don't think it's standard practice, but I do think it's a good idea regardless of what is currently being done in nixpkgs (nixpkgs is not some kind of designed engineering perfection, so in a sense it doesn't really matter what other modules do). You could be first to do things right! Imagine that!

Separate PR is fine.

The use-case of something like this would be to ideally even include a clickable link in the generated files themselves too such that if you open them you can immediately jump to the defining source in case you want to make modifications.

(I don't speak on behalf of the NixOS organisation, so whether or not you listen to me is completely optional. The only effect is that the number of future users of this module would likely be one less if it diverges too far from our our goals. )